### PR TITLE
Add FlaxCLIP

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

This PR adds the CLIP model in JAX/Flax.

### CODE REVIEW

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 1]
```diff
@@ -92,7 +92,8 @@ def convert_pytorch_state_dict_to_flax(pt_state_dict, flax_model):
             pt_tuple_key = pt_tuple_key[:-1] + ("embedding",)
         elif pt_tuple_key[-1] == "weight" and pt_tuple_key not in random_flax_state_dict:
             pt_tuple_key = pt_tuple_key[:-1] + ("kernel",)
-            pt_tensor = pt_tensor.T
+            is_conv = pt_tensor.ndim == 4  # hackish way of detecting conv layer
+            pt_tensor = pt_tensor.transpose(2, 3, 1, 0) if is_conv else pt_tensor.T
```
Comment (patil-suraj) [inline src/transformers/modeling_flax_pytorch_utils.py:96]: Flax's conv layer is channels's last, so we need to transpose weights

File: src/transformers/models/clip/modeling_flax_clip.py [hunk 2]
```diff
@@ -0,0 +1,1087 @@
...
+            rngs=rngs,
+        )
+
+
+class FlaxCLIPVisionPreTrainedModel(FlaxPreTrainedModel):
+    config_class = CLIPVisionConfig
+    module_class: nn.Module = None
+
+    def __init__(self, config: CLIPVisionConfig, seed: int = 0, dtype: jnp.dtype = jnp.float32, **kwargs):
+        input_shape = (1, config.image_size, config.image_size, 3)
```
Comment (patil-suraj) [inline src/transformers/models/clip/modeling_flax_clip.py:663]: using hardcoded shape here, because the `conv` layer's weight dim depend on the `image_size` and we need config to access the `image_size`

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 3]
```diff
@@ -92,7 +92,8 @@ def convert_pytorch_state_dict_to_flax(pt_state_dict, flax_model):
             pt_tuple_key = pt_tuple_key[:-1] + ("embedding",)
         elif pt_tuple_key[-1] == "weight" and pt_tuple_key not in random_flax_state_dict:
             pt_tuple_key = pt_tuple_key[:-1] + ("kernel",)
-            pt_tensor = pt_tensor.T
+            is_conv = pt_tensor.ndim == 4  # hackish way of detecting conv layer
```
Comment (sgugger) [inline src/transformers/modeling_flax_pytorch_utils.py:95]: Can we get something from the name for this?

File: src/transformers/models/clip/modeling_flax_clip.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/models/clip/modeling_flax_clip.py:663]: How do the conv layer require the image size in flax but not in PyTorch? In PyTorch it's only used for the Embeddings.

**Review by sgugger — APPROVED** [review]
Looking great! Could you add some examples in the docstrings on how to use it?

File: src/transformers/models/clip/modeling_flax_clip.py (hunk 2 above)
Comment (patil-suraj) [reply src/transformers/models/clip/modeling_flax_clip.py:663]: Aah, sorry, I said it the wrong way.

What I mean to say is, in flax we init the model by doing a forward pass so we need to pass an example, which is created using this `input_shape`. Here the `num_patches` depend on the `image_size` and `patch_size`, and `num_patches` is used to init the embedding layer. 

The default `input_shape` is the shape used by pre-trained model which would be (224, 224, 3) but in our tests, we use different `image_size` but we don't pass `input_shape` in tests, so in this case, it results in in-correct `num_patches` and there is shape mismatch when adding the embeddings because `input_shape` and current `image_size` in the test are different.

Maybe we should pass `input_shape` when initing the model in tests.
cc @patrickvonplaten

File: src/transformers/models/clip/modeling_flax_clip.py [hunk 4]
```diff
@@ -0,0 +1,1087 @@
...
+        >>> processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")
+
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+
+        >>> inputs = processor(text=["a photo of a cat", "a photo of a dog"], images=image, return_tensors="np", padding=True)
+
+        >>> outputs = model(**inputs)
+        >>> logits_per_image = outputs.logits_per_image # this is the image-text similarity score
+        >>> probs = jax.nn.softmax(logits_per_image, axis=1) # we can take the softmax to get the label probabilities
```
Comment (patil-suraj) [inline src/transformers/models/clip/modeling_flax_clip.py:1083]: @sgugger I've added an example in each `FlaxCLIPPreTrainedModel`'s docstring.

File: src/transformers/models/clip/modeling_flax_clip.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/models/clip/modeling_flax_clip.py:663]: Aaaaah! I understand! Then I think you should be using an init argument for that, that defaults to `config.image_size`, so it's easy to change.

Though all config arguments are easy to change so it works fine as it is too. Happy to see what @patrickvonplaten and @LysandreJik think is best here!

File: src/transformers/models/clip/modeling_flax_clip.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/models/clip/modeling_flax_clip.py:1083]: Sorry I missed those!

File: src/transformers/modeling_flax_pytorch_utils.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_flax_pytorch_utils.py:95]: Ah that's unfortunate that we can't really do much here except matching by `.ndim` :-/ since both Conv and Linear are called `weight` in PT and `kernel` in Flax. 

Could we make it a bit more readable though by adding a new elif:

```python
elif pt_tuple_key[-1] == "weight" and and pt_tensor.ndim == 4 and pt_tuple_key not in random_flax_state_dict:
# conv layer
   ....
elif pt_tuple_key[-1] == "weight" and pt_tuple_key not in random_flax_state_dict:
# dense layer
   ....

File: src/transformers/modeling_flax_pytorch_utils.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_flax_pytorch_utils.py:96]: put those into the separete elif branches would be nice I think

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 5]
```diff
@@ -172,7 +173,8 @@ def load_flax_weights_in_pytorch_model(pt_model, flax_state):
         # rename flax weights to PyTorch format
         if flax_key_tuple[-1] == "kernel" and ".".join(flax_key_tuple) not in pt_model_dict:
             flax_key_tuple = flax_key_tuple[:-1] + ("weight",)
-            flax_tensor = flax_tensor.T
+            is_conv = flax_tensor.ndim == 4  # hackish way of detecting conv layer
+            flax_tensor = jnp.transpose(flax_tensor, (3, 2, 0, 1)) if is_conv else flax_tensor.T
```
Comment (patrickvonplaten) [inline src/transformers/modeling_flax_pytorch_utils.py:177]: same can we add a new `elif here to not add another `if` abstraction level?

File: src/transformers/models/clip/modeling_flax_clip.py [hunk 6]
```diff
@@ -0,0 +1,1074 @@
...
+
+        self.num_patches = (image_size // patch_size) ** 2
+        num_positions = self.num_patches + 1
+        self.position_embedding = nn.Embed(num_positions, embed_dim, embedding_init=jax.nn.initializers.normal())
+        self.position_ids = jnp.expand_dims(jnp.arange(0, num_positions, dtype="i4"), axis=0)
+
+    def __call__(self, pixel_values):
+        batch_size = pixel_values.shape[0]
+        patch_embeds = self.patch_embedding(pixel_values)
+        n, h, w, c = patch_embeds.shape
```
Comment (patrickvonplaten) [inline src/transformers/models/clip/modeling_flax_clip.py:220]: can we get better names here? Don't like those single letters

File: src/transformers/models/clip/modeling_flax_clip.py [hunk 7]
```diff
@@ -0,0 +1,1074 @@
...
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        last_hidden_state = encoder_outputs[0]
+        last_hidden_state = self.final_layer_norm(last_hidden_state)
+
+        # text_embeds.shape = [batch_size, n_ctx, transformer.width]
+        # take features from the eot embedding (eot_token is the highest number in each sequence)
```
Comment (patrickvonplaten) [inline src/transformers/models/clip/modeling_flax_clip.py:527]: What does `eot` stand for? was EOS meant?

File: src/transformers/models/clip/modeling_flax_clip.py [hunk 8]
```diff
@@ -0,0 +1,1074 @@
...
+        return_dict: bool = True,
+    ):
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        hidden_states = self.embeddings(pixel_values)
+        hidden_states = self.pre_layrnorm(hidden_states)
```
Comment (patrickvonplaten) [inline src/transformers/models/clip/modeling_flax_clip.py:566]: typo? `pre_layernorm`?

Comment (patrickvonplaten) [reply src/transformers/models/clip/modeling_flax_clip.py:566]: Is this wrong in PyTorch as well?

File: src/transformers/models/clip/modeling_flax_clip.py [hunk 9]
```diff
@@ -0,0 +1,1074 @@
...
+            output_hidden_states,
+            return_dict,
+            rngs=rngs,
+        )
+
+    def get_text_features(
+        self,
+        input_ids,
+        attention_mask=None,
+        position_ids=None,
```
Comment (patrickvonplaten) [inline src/transformers/models/clip/modeling_flax_clip.py:777]: why not add 

```
 not train,            
output_attentions,           
 output_hidden_states,            
return_dict,            
rngs=rngs,
``` 

here?

File: src/transformers/models/clip/modeling_flax_clip.py [hunk 10]
```diff
@@ -0,0 +1,1074 @@
...
+            pooled_output = text_outputs[1]
+            text_features = module.text_projection(pooled_output)
+            return text_features
+
+        return self.module.apply(
+            {"params": self.params},
+            jnp.array(input_ids, dtype="i4"),
+            jnp.array(attention_mask, dtype="i4"),
+            jnp.array(position_ids, dtype="i4"),
+            method=_get_features,
```
Comment (patrickvonplaten) [inline src/transformers/models/clip/modeling_flax_clip.py:816]: nice!

File: src/transformers/models/clip/modeling_flax_clip.py [hunk 11]
```diff
@@ -0,0 +1,1074 @@
...
+        def _get_features(module, pixel_values):
+            vision_outputs = module.vision_model(pixel_values=pixel_values)
+            pooled_output = vision_outputs[1]  # pooled_output
+            image_features = module.visual_projection(pooled_output)
+            return image_features
+
+        return self.module.apply(
+            {"params": self.params},
+            jnp.array(pixel_values, dtype=jnp.float32),
+            method=_get_features,
```
Comment (patrickvonplaten) [inline src/transformers/models/clip/modeling_flax_clip.py:843]: same I think we could provide `train`, `return_dict` here as well no?

File: tests/test_modeling_flax_clip.py [hunk 12]
```diff
@@ -0,0 +1,473 @@
...
+
+    @slow
+    def test_model_from_pretrained(self):
+        for model_class_name in self.all_model_classes:
+            model = model_class_name.from_pretrained("openai/clip-vit-base-patch32", from_pt=True)
+            outputs = model(input_ids=np.ones((1, 1)), pixel_values=np.ones((1, 3, 224, 224)))
+            self.assertIsNotNone(outputs)
+
+    @is_pt_flax_cross_test
+    def test_equivalence_pt_to_flax(self):
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_clip.py:388]: why do we have to overwrite the test?

File: tests/test_modeling_flax_clip.py [hunk 13]
```diff
@@ -0,0 +1,473 @@
...
+
+                with torch.no_grad():
+                    pt_outputs_loaded = pt_model_loaded(**pt_inputs).to_tuple()
+                pt_outputs_loaded = pt_outputs_loaded[1:]
+
+                self.assertEqual(
+                    len(fx_outputs), len(pt_outputs_loaded), "Output lengths differ between Flax and PyTorch"
+                )
+                for fx_output, pt_output in zip(fx_outputs[:4], pt_outputs_loaded[:4]):
+                    self.assert_almost_equals(fx_output, pt_output.numpy(), 4e-2)
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_clip.py:473]: Can we add a test for `get_image_features` as well (ideally a jitted test)? 
and for `get_text_features`?

File: src/transformers/models/clip/modeling_flax_clip.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/models/clip/modeling_flax_clip.py:659]: Agree with @sgugger, think we should provide an `input_shape` argument that defaults to `(1, config.image_size, config.image_size, 3)` if `None`. We also provide an `input_sahpe` to FlaxBert's `__init__(...)` function

File: src/transformers/models/clip/modeling_flax_clip.py (hunk 8 above)
Comment (patil-suraj) [reply src/transformers/models/clip/modeling_flax_clip.py:565]: yeah, unfortunately it's a typo in Pytorch :(

File: src/transformers/models/clip/modeling_flax_clip.py (hunk 11 above)
Comment (patil-suraj) [reply src/transformers/models/clip/modeling_flax_clip.py:869]: added the `train` and `rngs` args. These functions are meant to get the image/text features so we don't return all outputs (attentions, hidden_states) from these.

File: tests/test_modeling_flax_clip.py (hunk 12 above)
Comment (patil-suraj) [reply tests/test_modeling_flax_clip.py:424]: The `FlaxCLIPModel` returns nested outputs (vision model's output, text model's output),  which is not supported by this test.

File: src/transformers/models/clip/modeling_flax_clip.py (hunk 11 above)
Comment (patrickvonplaten) [reply src/transformers/models/clip/modeling_flax_clip.py:869]: ok!

File: tests/test_modeling_flax_clip.py (hunk 12 above)
Comment (patrickvonplaten) [reply tests/test_modeling_flax_clip.py:424]: ok! Can we add a short comment maybe?

**Review by patrickvonplaten — APPROVED** [review]
After adding some more tests and allowing an `input_shape` input to `__init__(...)` - this LGTM :-)

File: tests/test_modeling_flax_clip.py (hunk 13 above)
Comment (patil-suraj) [reply tests/test_modeling_flax_clip.py:509]: do you mean an equivalence test?

Comment (patrickvonplaten) [reply tests/test_modeling_flax_clip.py:509]: I meant a jitted test - for those functions or did you already add one?

**Comment by patil-suraj** [comment]
- Added jitted tests for `get_image_features` and `get_text_features`
- `__init__(...)` now takes `input_shape` as an arg, when `None` it's set using default values in `config`.

Merging!
